### PR TITLE
Fix some of gcc7 warnings

### PR DESCRIPTION
**Pull request by sovrasov**

### CODE REVIEW

File: modules/stereo/include/opencv2/stereo/matching.hpp [hunk 1]
```diff
@@ -423,7 +423,6 @@ namespace cv
             //preprocessing the cost volume in order to get it ready for aggregation
             void costGathering(const Mat &hammingDistanceCost, Mat &cost)
             {
-                CV_Assert(hammingDistanceCost.rows == hammingDistanceCost.rows);
```
Comment (mshabunin) [inline modules/stereo/include/opencv2/stereo/matching.hpp]: Shouldn't it be `rows == cols` assertion?

File: modules/stereo/test/test_block_matching.cpp [hunk 2]
```diff
@@ -97,7 +97,7 @@ void CV_BlockMatchingTest::run(int )
         ts->set_failed_test_info(cvtest::TS::FAIL_INVALID_TEST_DATA);
         return;
     }
-    if(image1.rows != image2.rows || image1.cols != image2.cols || gt.cols != gt.cols || gt.rows != gt.rows)
```
Comment (mshabunin) [inline modules/stereo/test/test_block_matching.cpp]: Shouldn't it be `gt.cols != image1.cols || gt.rows != image1.rows`?

File: modules/stereo/test/test_block_matching.cpp [hunk 3]
```diff
@@ -181,7 +181,7 @@ void CV_SGBlockMatchingTest::run(int )
...
```
Comment (mshabunin) [inline modules/stereo/test/test_block_matching.cpp]: Same as for line 100

File: modules/xfeatures2d/src/harris_lapace_detector.cpp [hunk 4]
```diff
@@ -4,6 +4,10 @@
 
 #include "precomp.hpp"
 
+#if defined __GNUC__
+#  pragma GCC diagnostic ignored "-Wunused-function"
```
Comment (mshabunin) [inline modules/xfeatures2d/src/harris_lapace_detector.cpp]: Maybe is it better to remove unused function?

Comment (sovrasov) [reply modules/xfeatures2d/src/harris_lapace_detector.cpp]: Fixed

**Comment by mshabunin** [comment]
Thanks!
:+1:

**Pull request merged**